### PR TITLE
feat(frontend): Muti charge type

### PR DESCRIPTION
**Pull request by zhouzilong2020**
This should follow #17
截屏2022-04-26 19 46 04

### CODE REVIEW

**Comment by d-bytebase** [comment]
I will let @LiuJi-Jim to review this frontend change.

**Comment by zhouzilong2020** [comment]
> We can hide "lease length" column if the user only selects ON DEMAND.
> 
> Instead of showing INF, we can display N/A

Now the data table use a dynamic head. 
1. Show the `Engine` column only when multiple engines are selected.

2. Show the `Lease Length` column when `Reserved` or multiple charge types are selected.

Done, PTAL;

**Review by tianzhou — APPROVED** [review]
LGTM

File: frontend/src/components/CostTable.vue [hunk 1]
```diff
@@ -180,8 +200,8 @@ const props = defineProps({
     default: [],
   },
   chargeType: {
-    type: String as PropType<ChargeType>,
-    default: "",
+    type: Object as PropType<ChargeType[]>,
```
Comment (LiuJi-Jim) [inline frontend/src/components/CostTable.vue:203]: `type: Array as PropType<ChargeType[]>`

File: frontend/src/components/CostTable.vue [hunk 2]
```diff
@@ -180,8 +200,8 @@ const props = defineProps({
...
+    default: [""],
```
Comment (LiuJi-Jim) [inline frontend/src/components/CostTable.vue:204]: `default: () => []`
or 
`default: () => [""]`
when a prop is defined as an Object or an Array, its `default` need to be a function to generate the default value.

File: frontend/src/components/CostTable.vue [hunk 3]
```diff
@@ -303,13 +324,14 @@ const refreshDataTable = () => {
           engineType: term.databaseEngine,
           commitment: { usd: term.commitmentUSD },
           hourly: { usd: term.hourlyUSD },
-          leaseLength: term.payload?.leaseContractLength,
+          leaseLength: term.payload?.leaseContractLength
+            ? term.payload?.leaseContractLength
+            : "N/A",
```
Comment (LiuJi-Jim) [inline frontend/src/components/CostTable.vue:329]: try `leaseLength: term.payload?.leaseContractLength ?? "N/A"`

File: frontend/src/components/CostTableMenu.vue [hunk 4]
```diff
@@ -116,8 +113,8 @@ const props = defineProps({
     default: [],
   },
   chargeType: {
-    type: String as PropType<ChargeType>,
-    default: "",
+    type: Object as PropType<ChargeType[]>,
+    default: [""],
```
Comment (LiuJi-Jim) [inline frontend/src/components/CostTableMenu.vue:117]: same here

File: frontend/src/components/CostTableMenu.vue [hunk 5]
```diff
@@ -161,7 +158,7 @@ const handleUpdateRegion = (val: any[]) => {
   emit("update-region", val);
 };
 
-const handleUpdateChargeType = (val: ChargeType) => {
+const handleUpdateChargeType = (val: any) => {
```
Comment (LiuJi-Jim) [inline frontend/src/components/CostTableMenu.vue:161]: Is it intentionally to use `any` here?

**Review by LiuJi-Jim — COMMENTED** [review]
LG overall, with some minor comments PTAL.

File: frontend/src/components/CostTableMenu.vue (hunk 5 above)
Comment (zhouzilong2020) [reply frontend/src/components/CostTableMenu.vue:161]: I just don't know how to deal with the type assertion. 
`Can not allocate type “(val: string[]) => void” to type “MaybeArray<(value: (string | number)[]) => void> | undefined”。`

**Review by LiuJi-Jim — APPROVED** [review]
LGTM

**Review by d-bytebase — APPROVED** [review]

File: frontend/src/components/CostTableMenu.vue (hunk 5 above)
Comment (LiuJi-Jim) [reply frontend/src/components/CostTableMenu.vue:161]: That's okay. Let's find a way later.

**Comment by d-bytebase** [comment]
merged conflict

**Pull request merged**